src/System.Runtime/tests/System/String.cs

```diff
@@ -88,93 +61,73 @@ public static void TestCtorCharPtrIntInt()
-
-        s = new String('a', 0);
-        Assert.Equal(String.Empty, s);
-
-        s = new String('a', 1);
-        Assert.Equal("a", s);
-
-        s = new String('a', 2);
-        Assert.Equal("aa", s);
-
-        s = new String('a', 3);
-        Assert.Equal("aaa", s);
-
-        s = new String('a', 4);
-        Assert.Equal("aaaa", s);
-
-        s = new String('a', 5);
-        Assert.Equal("aaaaa", s);
-
-        s = new String('a', 6);
-        Assert.Equal("aaaaaa", s);
-
-        s = new String('a', 7);
-        Assert.Equal("aaaaaaa", s);
-
-        s = new String('a', 8);
-        Assert.Equal("aaaaaaaa", s);
-
-        s = new String('a', 9);
-        Assert.Equal("aaaaaaaaa", s);
-
-        Assert.Throws<ArgumentOutOfRangeException>(() => s = new String('a', -1));
+        String s = new String(c, count);
+        Assert.Equal(expected, s);
     }
 
     [Fact]
-    public static void TestCtorCharArray()
+    public static void TestCtorCharIntInvalid()
     {
-        String s;
-
-        char[] c = { 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h' };
-        s = new String(c);
-        Assert.Equal("abcdefgh", s);
-
-        s = new String((char[])null);
-        Assert.Equal(String.Empty, s);
+        Assert.Throws<ArgumentOutOfRangeException>(() => new String('a', -1));
     }
 
-    [Fact]
-    public static void TestCtorCharArrayIntInt()
+    [Theory]
+    [InlineData(new char[] { }, 0, 0, "")]
+    [InlineData((char[])null, 0, 0, "")]
+
+    [InlineData(new char[] { 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h' }, 0, 0, "")]
+    [InlineData(new char[] { 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h' }, 0, 3, "abc")]
+    [InlineData(new char[] { 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h' }, 2, 3, "cde")]
+    [InlineData(new char[] { 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h' }, 0, 8, "abcdefgh")]
+    [InlineData(new char[] { 'a', 'b', 'c', 'd', '\0', 'e', 'f', 'g', 'h' }, 0, 9, "abcd\0efgh")]
+    [InlineData(new char[] { 'П', 'Р', 'И', 'В', 'Е', 'Т' }, 0, 6, "ПРИВЕТ")]
+    public static void TestCtorCharArrar(char[] c, int startIndex, int length, string expected)
     {
-        String s;
-
-        char[] c = { 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h' };
-        s = new String(c, 2, 3);
-        Assert.Equal("cde", s);
-
-        s = new String(c, 0, 8);
-        Assert.Equal("abcdefgh", s);
-
-        s = new String(c, 0, 0);
-        Assert.Equal("", s);
-
-        Assert.Throws<ArgumentOutOfRangeException>(() => s = new String(c, 0, 9));
-
-        Assert.Throws<ArgumentOutOfRangeException>(() => s = new String(c, -1, 1));
-
-        Assert.Throws<ArgumentOutOfRangeException>(() => s = new String(c, 5, -1));
-
-        Assert.Throws<ArgumentNullException>(() => s = new String((char[])null, 0, 0));
-
-        Assert.Throws<ArgumentNullException>(() => s = new String((char[])null, 3, 9));
+        if (c == null)
+        {
+            Assert.Equal(expected, new String(c));
+            return;
+        }
```

Comment:
Only one of the inputs is null.  Rather than running this for every test case, it can be separated out into its own Fact that tests the ctor with null.